// @devon4node/common/serializer/interceptors/class-serializer.interceptor.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClassSerializerInterceptor = void 0;
const operators_1 = require("rxjs/operators");
const common_1 = require("@nestjs/common");
const load_package_util_1 = require("@nestjs/common/utils/load-package.util");
const shared_utils_1 = require("@nestjs/common/utils/shared.utils");
const class_serializer_constants_1 = require("@nestjs/common/serializer/class-serializer.constants");
const defaults_1 = require("../defaults");
let classTransformer = {};
const REFLECTOR = 'Reflector';
let ClassSerializerInterceptor = class ClassSerializerInterceptor {
    constructor(reflector) {
        this.reflector = reflector;
        classTransformer = load_package_util_1.loadPackage('class-transformer', 'ClassSerializerInterceptor', () => require('class-transformer'));
        require('class-transformer');
    }
    intercept(context, next) {
        const options = this.getContextOptions(context);
        const type = this.getType(context);
        return next
            .handle()
            .pipe(operators_1.map((res) => this.serialize(res, options, type)));
    }
    serialize(response, options, type) {
        if (!response) {
            return response;
        }
        if (type) {
            if (response.data) {
                const data = this.transformToClass(type, response.data, options);
                response.data = this.transformToPlain(data, options);
                return response;
            }
            return classTransformer.classToPlain(this.transformToClass(type, response, options), options);
        }
        else {
            const isArray = Array.isArray(response);
            if (!shared_utils_1.isObject(response) && !isArray) {
                return response;
            }
            return isArray
                ? response.map(item => this.transformToPlain(item, options))
                : this.transformToPlain(response, options);
        }
    }
    transformToPlain(plainOrClass, options) {
        return plainOrClass && plainOrClass.constructor !== Object
            ? classTransformer.classToPlain(plainOrClass, options)
            : plainOrClass;
    }
    transformToClass(classType, plainOrClass, options) {
        return classTransformer.plainToClass(classType, plainOrClass, options);
    }
    getContextOptions(context) {
        return this.reflectSerializeMetadata(context.getHandler()) || this.reflectSerializeMetadata(context.getClass());
    }
    getType(context) {
        return (this.reflector.get(defaults_1.CRUD_TYPE_ID, context.getHandler()) || this.reflector.get(defaults_1.CRUD_TYPE_ID, context.getClass()));
    }
    reflectSerializeMetadata(obj) {
        return this.reflector.get(class_serializer_constants_1.CLASS_SERIALIZER_OPTIONS, obj);
    }
};
ClassSerializerInterceptor = __decorate([
    common_1.Injectable(),
    __param(0, common_1.Inject(REFLECTOR)),
    __metadata("design:paramtypes", [Object])
], ClassSerializerInterceptor);
exports.ClassSerializerInterceptor = ClassSerializerInterceptor;
